components/salsa-2022-macros/src/input.rs

```diff
@@ -198,11 +217,86 @@ impl InputStruct {
             }
         }
+    }
+
+    /// generate builder struct
+    fn input_builder(&self) -> (syn::ItemStruct, syn::ItemImpl) {
+        let ident = self.id_ident();
+        let struct_name = self.builder_name();
+        let field_names = self.all_field_names();
+        let field_tys: Vec<_> = self.all_field_tys();
+        let db_dyn_ty = self.db_dyn_ty();
+        let jar_ty = self.jar_ty();
+        let input_index = self.input_index();
+        let field_indices = self.all_field_indices();
+
+        let buidler: syn::ItemStruct = parse_quote! {
+            struct #struct_name{
+                #(#field_names: Option<#field_tys>,)*
+                durability: Option<salsa::durability::Durability>,
+            }
+        };
+
+        let build_fn: syn::ImplItemMethod = if self.0.is_isingleton() {
+            parse_quote! {
+                fn build(&mut self, __db: &#db_dyn_ty) -> std::result::Result<#ident, std::boxed::Box<dyn std::error::Error>> {
+                    let durability = self.durability.ok_or_else(|| "durability not provided")?;
 
-        // }
+                    let (__jar, __runtime) = <_ as salsa::storage::HasJar<#jar_ty>>::jar(__db);
+                    let __ingredients = <#jar_ty as salsa::storage::HasIngredientsFor< #ident >>::ingredient(__jar);
+                    let __id = __ingredients.#input_index.new_singleton_input(__runtime);
+
+                    #(
+                        __ingredients.#field_indices.store_new(__runtime, __id, self.#field_names.clone().ok_or_else(||"field not provided")?, durability);
+                    )*
+                    std::result::Result::Ok(__id)
+                }
+            }
+        } else {
+            parse_quote! {
+                fn build(&mut self, __db: &#db_dyn_ty) -> std::result::Result<#ident, std::boxed::Box<dyn std::error::Error>> {
+                    let durability = self.durability.ok_or_else(|| "durability not provided")?;
+
+                    let (__jar, __runtime) = <_ as salsa::storage::HasJar<#jar_ty>>::jar(__db);
+                    let __ingredients = <#jar_ty as salsa::storage::HasIngredientsFor< #ident >>::ingredient(__jar);
+                    let __id =__ingredients.#input_index.new_input(__runtime);
+                    #(
+                        __ingredients.#field_indices.store_new(__runtime, __id, self.#field_names.clone().ok_or_else(||"field not provided")?, durability);
+                    )*
+                    std::result::Result::Ok(__id)
+                }
+            }
+        };
+
+        let impls: syn::ItemImpl = {
+            parse_quote! {
+                impl #struct_name {
+                    fn with_durability(&mut self, durability: salsa::durability::Durability) -> &mut Self {
+                        self.durability = Some(durability);
+                        self
+                    }
+
+                    fn with_fields(&mut self, #(#field_names: #field_tys,)*) -> &mut Self {
```

Comment:
`with_values` would be a better name in my opinion, but the current is fine too.

Reply:
Hmmn. `with_fields` makes it obvious to the user what they're about to set. Wouldn't `with_values` be a little ambiguous? I really don't know.

Reply:
I think of a field as something which has a value, a durability and maybe other things associated with it, which is why `with_values` is better in my opinion as it communicates that you're setting just the value of the field. If I saw `with_fields` my first thought would be that it takes a struct or tuple which contains everything about a field. Maybe someone else could comment on what is the better option, but I think both are good?